### PR TITLE
add detach for Variable in dygraph, test=develop

### PR DESCRIPTION
**Pull request by JepsonWong**
Add detach for Variable in dygraph.
This interface returns a new Variable (stop_gradient=True), detached from the current graph.

### CODE REVIEW

File: python/paddle/fluid/framework.py [hunk 1]
```diff
@@ -34,6 +34,8 @@
 from . import core
 from . import unique_name
 
+import paddle
```
Comment (JiabinYang) [inline python/paddle/fluid/framework.py]: DO NOT import paddle here, this may cause circle deps

File: python/paddle/fluid/framework.py [hunk 2]
```diff
@@ -543,6 +545,16 @@ def __init__(self,
             self._stop_gradient = stop_gradient
             self.is_data = is_data
 
+    def detach(self):
+        new_var = self._cloneVar()
```
Comment (JiabinYang) [inline python/paddle/fluid/framework.py]: use `in_dygraph_mode()` to make sure this will not be used in static graph mode, and throw error when user use it. What's more, add Doc
for this API just like other apis

File: python/paddle/fluid/tests/unittests/test_detach.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                initializer=fluid.initializer.Constant(9.0))
+            fc2_b_param_attrs = fluid.ParamAttr(
+                initializer=fluid.initializer.Constant(10.0))
+            fc2 = FC("fc",
+                     1,
+                     num_flatten_dims=1,
+                     param_attr=fc2_w_param_attrs,
+                     bias_attr=fc2_b_param_attrs)
+            data = to_variable(data)
+            conv = fc(data)
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/test_detach.py]: why do u call this conv? change name to fc?

File: python/paddle/fluid/tests/unittests/test_detach.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            dim = dim * index_shape
+        return dim
+
+    def test_NoDetachMulti_DetachMulti(self):
+        array_no_detach_multi = self.no_detach_multi()
+        array_detach_multi = self.detach_multi()
+
+        array_no_detach_multi_dim = self.get_Dim(array_no_detach_multi)
+
+        array_no_detach_multi = array_no_detach_multi.reshape(
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/test_detach.py]: no need to do this, why not get numpy value and use np.array_equalI() to assert the value equality

**Review by JiabinYang — COMMENTED** [review]
Some comments

File: python/paddle/fluid/framework.py (hunk 1 above)
Comment (JepsonWong) [reply python/paddle/fluid/framework.py]: done

File: python/paddle/fluid/framework.py (hunk 2 above)
Comment (JepsonWong) [reply python/paddle/fluid/framework.py]: done

File: python/paddle/fluid/tests/unittests/test_detach.py (hunk 3 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/test_detach.py]: done

File: python/paddle/fluid/tests/unittests/test_detach.py (hunk 4 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/test_detach.py]: done

File: python/paddle/fluid/tests/unittests/test_detach.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        array_detach_multi = self.detach_multi()
+        assert np.array_equal(array_no_detach_single, array_detach_multi)
+
+    def test_detach_exception(self):
+        x = fluid.layers.data(name="a", shape=[3, 4], dtype='float32')
+        y = fluid.layers.fc(input=x, size=10, bias_attr=True)
+        try:
+            y_detach = y.detach()
+        except Exception as e:
+            print(e)
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/test_detach.py]: This print seems has no effect for ut, better option is raise an specific exception and catch it here, or do some string match to make sure it raises a predictable exception

**Review by JiabinYang — COMMENTED** [review]
some comments

File: python/paddle/fluid/tests/unittests/test_detach.py (hunk 5 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/test_detach.py]: done

**Review by JiabinYang — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]
LGTM for framework.py
